### ISSUE TITLE
RFC: Temporarily fork and develop a separate native version

### ISSUE THREAD

**Issue by kitten** [styled-components/styled-components#1617]
## Summary

Let's start a separate "styled-native-next" fork that'd be used to:

- Rewrite large parts of our codebase (especially the ones providing legacy hacks and support)
- Provide the opportunity to split out a "core" package which can be used for the web library eventually from this rewrite
- Improve React Native support and performance
- Provide a breeding ground for our ISTF/Sweetsour effort

## Motivation

Currently there's two problems that we're aiming to solve in the future. Our codebase's organisation and some bundlesize and performance concerns.

Most of these can be addressed by larger refactors, which are currently hard to accomplish and navigate. Instead of revamping the entire codebase in one fell swoop, we can opt to develop a "styled-native-next" package.

This package actually has another motivation entirely: Provide a more solid and performant styled-components version for React Native. It would address the problems around forwarding interpolations to the final styles that reach React Native, which would improve our support for the `Animated` API.

This is all possible and combined with out Sweetsour/ISTF effort (See: https://github.com/philpl/sweetsour). Which is easier to start with on React Native, since it won't need any CSS rule flattening or autoprefixing — only shorthands will need to be implemented.

Also, at the end of this refactor we'd hopefully be in a position to create a "core" package that can be used by the web library, which can eventually lead to a "web", "native", and "core" repo setup.

## Plan

- Fork the current library
- Remove web-specific bits
- Integrate the sweetsour-parser, optimally with a babel plugin
- Write a shorthand transformer, in parallel to other tasks
- Rewrite core parts of the library, like the interpolation handling and StyledComponent file
- Split generated ISTF styles into static, dynamic, and animated
- Segregate into a core/native package structure
- Reintegrate core package and refactors into a web shard of the library
- Combine new code into current repo

## Champions

- Myself (@kitten)
- Jani Eväkallio (@jevakallio) has helped to form this plan and will potentially be able to help out a little
- Evan Scott (@probablyup) is motivated in finding parts that can be refactored into the babel plugin and other parts of the source that can be rewritten for bundle size benefits

**Comment by Noitidart** [comment]
Thank you team!!!

If possible, it would be awesome if as much as possible used `StyleSheet.create`. That is a huge perf boost and one of the major reasons why we had to pull out styled-components in our RN project (i fought hard to keep it! :( i lost out). I know `StyleSheet.create` can't be used with Animated API, so this is aside from that please.

**Comment by kitten** [comment]
@Noitidart We’re already utilising it and are not planning to give it up 👍 It doesn’t actually provide a huge perf boost as per my research (yet) but it’s definitely important. animated support is also on my list of must-haves

**Comment by Noitidart** [comment]
Oh wow no way! That's awesome thanks @kitten!

**Comment by kristerkari** [comment]
This is a very interesting thing for a better React Native styling support!

I'm currently using [css-to-react-native](https://github.com/styled-components/css-to-react-native) to parse CSS in my own project [React Native CSS modules](https://github.com/kristerkari/react-native-css-modules).

I would be really interested in being kept in the loop, or even helping out with the CSS parser stuff. I feel like there is still quite a lot of things that we could add support for when it comes to styling in React Native.

**Comment by BraianMendes** [comment]
I loved this, will be a great help. I really like the styled-components and I missed something like that for native.

**Comment by Sharcoux** [comment]
Hey guys, I've been working on exactly this in [rn-css](https://github.com/Sharcoux/rn-css). I'm using it successfully in production in a very complex project. Maybe it's worth taking this as a starting point instead of adapting the code of styled-components which seems like a more complex work to do... Please give me your thoughts.

**Comment by Sharcoux** [comment]
Hum, so, what do you think guys?

**Comment by Sharcoux** [comment]
@kitten If you want to create styled-native-next or improve rn-css, please contact me. I'll be more than happy to help.

**Comment by quantizor** [comment]
Closing — this issue is stale (2+ years without resolution) and relates to an older version or has been superseded by v6.x changes. If this is still reproducible on the latest v6.x release, please open a new issue with a minimal reproduction.
